openmpi: ensure that shmem segments for co-located jobs don't conflict

garlick opened the pull request:
Problem: shared memory segment naming for the "vader" BTL may collide
when running multiple jobs per node.

This is because the path is constructed from
  /dev/shm/vader_segment.hostname.uid.jobid.rank

and the flux pmix plugin in openmpi always sets the jobid to 0
because FLUX_JOB_ID in f58 format is converted to a 32-bit integer
using strtol().  This is a bug in the plugin, but a proper fix may
be elusive given the integer size mismatch of IDs.

Work around the plugin bug by changing
  OMPI_MCA_btl_vader_backing_directory
from default of /dev/shm to $FLUX_JOB_TMPDIR, which the shell tmpdir
plugin will have set to a job-unique directory.

Fixes #3649

This is based on top of #3661 (FLUX_JOB_TMPDIR), which should be merged first.

grondo commented:
Honestly asking, what is the impact of use a possibly normal filesystem backed tmpdir for these segments instead of `/dev/shm`?

garlick commented:
According to [kernel docs](https://www.kernel.org/doc/gorman/html/understand/understand015.html) `/dev/shm` is just a regular tmpfs, so on our systems where `/tmp` is also a tmpfs, no difference?  Maybe some difference on systems where `/tmp` is a disk-backed file system?  Not sure.

FWIW, here is the ompi commit that moved the shared memory segments to `/dev/shm`.  I guess we are subverting this change, which was done for _some_ reason, performance or otherwise.

```
commit 47fd2313abc1cd78f42488ac04384c16ad5703bc
Author: Nathan Hjelm <hjelmn@lanl.gov>
Date:   Wed Dec 13 11:10:18 2017 -0700

    btl/vader: move backing files into /dev/shm on Linux
    
    This commit moves the backing files to /dev/shm to avoid limitations
    that may be set on /tmp. The files are registered with pmix to ensure
    they are cleaned up after an erroneous exit.
    
    Signed-off-by: Nathan Hjelm <hjelmn@lanl.gov>
    (cherry picked from commit 48101278160672317ade352365592f56ef3b8977)
```

grondo commented:
Well since this is just in a `.lua` plugin, sites can override this easily, so I don't think it is a problem. Maybe add a comment in the plugin?

garlick commented:
Ok, will do.

garlick commented:
Added comment
```lua
-- Note: this plugin changes the path for files that openmpi usually shares
-- using mmap (MAP_SHARED) from /dev/shm (tmpfs) to /tmp (maybe tmpfs).
-- Performance may be affected if /tmp is provided by a disk-backed file system.
-- This plugin may be overridden by loading another plugin with the same name.
-- See flux-shell(1).
```

and rebased on master now that FLUX_JOB_TMPDIR was merged.

grondo commented:
> This plugin may be overridden by loading another plugin with the same name. See flux-shell(1).

Actually, the `.lua` plugins can be considered config files and may be edited or replaced directly. They are not replaced by a plugin of the same name since they're loaded from the shell `initrc.lua`, and therefore callbacks registered from Lua 'plugins' are not associated with any plugin "name". (Sorry for the confusion here!)

garlick commented:
Oh!  Sorry, will fix.

grondo reviewed (approved):
LGTM, thanks!

garlick commented:
Thanks!  Setting MWP.
